utils/generator.py

```diff
@@ -0,0 +1,544 @@
+                            "required": False,
+                            "positional": True,
+                        },
+                    ]
+                    k["is_single_field"] = True
+                else:
+                    # special handling for multi-key dicts with field keys
+                    k["args"] = [
+                        {
+                            "name": "_fields",
+                            "type": f"Optional[Mapping[{key_type}, {value_type}]]",
+                            "doc": [
+                                ":arg _fields: A dictionary of fields with their values."
+                            ],
+                            "required": False,
+                            "positional": True,
+                        },
+                    ]
+                    k["is_multi_field"] = True
+            else:
+                raise RuntimeError(f"Cannot generate code for type {p['type']}")
+
+        else:
+            raise RuntimeError(f"Cannot generate code for type {p['type']}")
+        return k
+
+    def interface_to_python_class(self, interface, interfaces):
+        """Return a dictionary with template data necessary to render an
+        interface a Python class.
+
+        This is used to render interfaces that are referenced by container
+        classes. The current list of rendered interfaces is passed as a second
+        argument to allow this method to add more interfaces to it as they are
+        discovered.
+
+        The returned format is as follows:
+
+        ```python
+        {
+            "name": "the class name to use for the interface class",
+            "parent": "the parent class name",
+            "args": [ # a Python description of each class attribute
+                "name": "the name of the attribute",
+                "type": "the Python type hint for the attribute",
+                "doc": ["formatted lines of documentation to add to class docstring"],
+                "required": bool,
+                "positional": bool,
+            ],
+        }
+        ```
+        """
+        type_ = schema.find_type(interface)
+        if type_["kind"] != "interface":
+            raise RuntimeError(f"Type {interface} is not an interface")
+        k = {"name": interface, "args": []}
+        while True:
+            for arg in type_["properties"]:
+                schema.add_attribute(k, arg, for_types_py=True)
+            if "inherits" in type_ and "type" in type_["inherits"]:
+                if "parent" not in k:
+                    k["parent"] = type_["inherits"]["type"]["name"]
+                if type_["inherits"]["type"]["name"] not in interfaces:
+                    interfaces.append(type_["inherits"]["type"]["name"])
+                type_ = schema.find_type(
+                    type_["inherits"]["type"]["name"],
+                    type_["inherits"]["type"]["namespace"],
+                )
+            else:
+                break
```

Comment:
nit: I find placing the break up-front make the code easier to read.

```suggestion
        while True:
            if "inherits" not in type_ or "type" not in type_["inherits"]:
                break
                
            for arg in type_["properties"]:
                schema.add_attribute(k, arg, for_types_py=True)
            if "parent" not in k:
                k["parent"] = type_["inherits"]["type"]["name"]
            if type_["inherits"]["type"]["name"] not in interfaces:
                interfaces.append(type_["inherits"]["type"]["name"])
            type_ = schema.find_type(
                type_["inherits"]["type"]["name"],
                type_["inherits"]["type"]["namespace"],
            )
```